apollo-core/src/main/java/com/spotify/apollo/core/ServiceImpl.java

```diff
@@ -232,14 +236,20 @@ InstanceImpl initInstance(
         shutdownRequested, stopped);
   }
 
-  Set<ApolloModule> discoverAllModules() {
-    final Set<ApolloModule> allModules;
+  Set<ApolloModule> discoverAllModules(
+      Set<ApolloModule> extraModules) {
+
+    Set<ApolloModule> allModules =
+        Sets.union(
+            modules,
+            extraModules.stream()
+                .filter(m -> modules.stream().noneMatch(am -> m.getId().equals(am.getId())))
+                .collect(Collectors.toSet()));
 
     if (moduleDiscovery) {
-      allModules = Sets.union(modules, ImmutableSet.copyOf(ServiceLoader.load(ApolloModule.class)));
-    } else {
-      allModules = modules;
+      return Sets.union(allModules, ImmutableSet.copyOf(ServiceLoader.load(ApolloModule.class)));
```

Comment:
We're not doing the deduping against the modules loaded visa ServiceLoader, but I guess that's ok :)